### PR TITLE
BOOKKEEPER-588 SSL Support

### PR DESCRIPTION
**Pull request by eolivelli**

### CODE REVIEW

**Comment by jvrao** [comment]
Kishore is reviewing it. I hope he can make this mornings call. He is on vacation and supposed to be back today.

File: bookkeeper-server/nb-configuration.xml [hunk 1]
```diff
@@ -1,18 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (kishorekasi) [inline bookkeeper-server/nb-configuration.xml]: remove this netbeans relevant config file?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/AuthProviderFactoryFactory.java [hunk 2]
```diff
@@ -25,10 +25,6 @@
 import org.apache.bookkeeper.client.BKException;
 import org.apache.bookkeeper.conf.ClientConfiguration;
 import org.apache.bookkeeper.conf.ServerConfiguration;
-<<<<<<< HEAD
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/AuthProviderFactoryFactory.java]: Fix merge issues

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java [hunk 3]
```diff
@@ -59,11 +59,7 @@
          *                   is complete.
          */
         BookieAuthProvider newProvider(BookieConnectionPeer connection,
-<<<<<<< HEAD
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java]: Same here. Fix the merge issues in this file too

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java [hunk 4]
```diff
@@ -284,4 +289,46 @@ public AbstractConfiguration setClientAuthProviderFactoryClass(
...
+     * @return the client authentication provider factory class name or null.
+     */
+    public String getSslEnabledCipherSuites() {
+        return getString(SSL_ENABLED_CIPHER_SUITES, null);
+    }
+
+    /**
+     * Set the list of enabled SSL protocols. leave blank not to override default JDK list
+     *
+     * @param list comma separated list of enabled SSL cipher suites
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: fix comment?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java [hunk 5]
```diff
@@ -284,4 +289,46 @@ public AbstractConfiguration setClientAuthProviderFactoryClass(
...
+    public AbstractConfiguration setSslEnabledCipherSuites(
+            String list) {
+        setProperty(SSL_ENABLED_CIPHER_SUITES, list);
+        return this;
+    }
+
+    /**
+     * Get the list of enabled SSL cipher suites
+     *
+     * @return the client authentication provider factory class name or null.
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: fix comment

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java [hunk 6]
```diff
@@ -284,4 +289,46 @@ public AbstractConfiguration setClientAuthProviderFactoryClass(
...
+    public AbstractConfiguration setSslEnabledProtocols(
+            String list) {
+        setProperty(SSL_ENABLED_PROTOCOLS, list);
+        return this;
+    }
+
+    /**
+     * Get the list of enabled SSL protocols
+     *
+     * @return the client authentication provider factory class name or null.
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: fix comment

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 7]
```diff
@@ -101,6 +101,31 @@
...
+     */
+    protected final static String VERIFY_SSL_CERTIFICATES = "verifySSLCertificates";
+
+    /**
+     * Enable client authentication using an SSL certificate
+     */
+    protected final static String CLIENT_SSL_AUTHENTICATION = "clientSSLAuthentication";
+
+    /**
+     * Key store which holds the client-side SSL certificate. The certificate must be in PKCS12 format
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: Any reason why are assuming only pkcs#12 format. Why not JKS or pkcs#8?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 8]
```diff
@@ -882,6 +907,112 @@ public ClientConfiguration setBookieErrorThresholdPerInterval(long thresholdPerI
...
+    * Whether the client validates SSL certificates. It defaults to false
+    * @return whether the client will verify certificates or not.
+    */
+    public boolean getVerifySSLCertificates() {
+        return getBoolean(VERIFY_SSL_CERTIFICATES, false);
+    }
+
+    /**
+    * Specify whether the client validates SSL certificates
+    * @return whether the client will verify certificates or not.
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: fix comment

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 9]
```diff
@@ -882,6 +907,112 @@ public ClientConfiguration setBookieErrorThresholdPerInterval(long thresholdPerI
...
+    public ClientConfiguration setVerifySSLCertificates(boolean value) {
+        setProperty(VERIFY_SSL_CERTIFICATES, value);
+        return this;
+    }
+
+
+    /**
+      * Specify whether the client will send an SSL certificate on TLS-handshake
+      * @param Whether to send a certificate or not
+      * @return the server configuration
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: fix comment "server" -> "client"

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 10]
```diff
@@ -882,6 +907,112 @@ public ClientConfiguration setBookieErrorThresholdPerInterval(long thresholdPerI
...
+         setProperty(CLIENT_SSL_AUTHENTICATION, enabled);
+         return this;
+     }
+
+     /**
+      * Whether the client will send an SSL certificate on TLS-handshake
+      * @see #setClientSSLAuthentication(boolean)
+      * @return whether SSL is enabled on the bookie or not.
+      */
+     public boolean isClientSSLAuthentication() {
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: Reword function name or keep it consistent with the above convention of "set" or "get"

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 11]
```diff
@@ -101,6 +101,31 @@
...
+
+    /**
+     * Enable client authentication using an SSL certificate
+     */
+    protected final static String CLIENT_SSL_AUTHENTICATION = "clientSSLAuthentication";
+
+    /**
+     * Key store which holds the client-side SSL certificate. The certificate must be in PKCS12 format
+     */
+    protected final static String CLIENT_SSL_KEY_STORE = "clientSSLKeyStore";
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: For local CA signed certificates, should we also have an option to provide the trustStore along with client/bookie keyStore?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java (hunk 7 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: This comes from the original patch from @ivankelly. In production I usually use PKCS12. I can make a new option for the format. Do you think it wild be useful?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java [hunk 12]
```diff
@@ -70,6 +70,11 @@
     // Client auth provider factory class name. It must be configured on Bookies to for the Auditor
     protected final static String CLIENT_AUTH_PROVIDER_FACTORY_CLASS = "clientAuthProviderFactoryClass";
 
+    // Common SSL configuration
+    protected final static String SSL_ENABLED_CIPHER_SUITES = "sslEnabledCipherSuites";
+    protected final static String SSL_ENABLED_PROTOCOLS = "sslEnabledProtocols";
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: Please provide more information in the comment about what protocols means here. Do you mean protocol version here?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+            engine.setEnabledProtocols(protocols);
+        }
+        if (enabledCipherSuites != null && !enabledCipherSuites.isEmpty()) {
+            String[] ciphers = enabledCipherSuites.split(",");
+            engine.setEnabledCipherSuites(ciphers);
+        }
+        return engine;
+    }
+
+    private TrustManager[] getTrustManagers() {
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java]: To keep more secure and enable both client/bookie trust only certificates from trusted issuers, should we use TrustStore configuration and create a trust manager from the provided truststore?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    public SSLContext getContext() {
+        return ctx;
+    }
+
+    public SSLEngine getEngine() {
+        SSLEngine engine = ctx.createSSLEngine();
+        engine.setUseClientMode(isClient);
+        if (!isClient) {
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java]: Allowing both client and server to authenticate itself will enabled MutualTLS

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java [hunk 15]
```diff
@@ -79,16 +75,19 @@ default void close() {}
     }
 
     /**
+     * Callback to let the provider know that the underlying protocol is changed.
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java]: Is there a specific action that is needed as part of this callback in AuthProvider?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ConnectionPeer.java [hunk 16]
```diff
@@ -60,4 +60,10 @@
      * @see #getAuthorizedId()
      */
     public void setAuthorizedId(BookKeeperPrincipal principal);
+
+    /**
+     * This flag returns two if a 'secure' channel in use, like SSL
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ConnectionPeer.java]: fix comment. "two" -> "true"

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java [hunk 17]
```diff
@@ -1712,4 +1716,74 @@ public void setBookieAuthProviderFactoryClass(String factoryClass) {
...
+     public String getSSLKeyStore() {
+         return getString(SSL_KEY_STORE, "nonexistent_keystore");
+     }
+
+     /**
+      * Set the password to decrypt the configured SSL keystore.
+      * @see #setSSLKeyStore(String)
+      * @return the server configuration
+      */
+     public ServerConfiguration setSSLKeyStorePassword(String keyStorePassword) {
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java]: Will it be useful to have the keyStore encryption password to be store in a file and this config parameter to accept the path?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java [hunk 18]
```diff
@@ -34,7 +34,6 @@
 import org.apache.bookkeeper.conf.ClientConfiguration;
```
Comment (kishorekasi) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java]: As identified test cases are missing. Adding few more test cases to the list -

1. Multiple bookkeeper clients (a few with ssl enabled with multiple bookies (all ssl enabled)
2. Multiple bookkeeper clients (a few with ssl enabled with multiple bookies (a few with ssl disabled)

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java [hunk 19]
```diff
@@ -344,7 +335,8 @@ void connectIfNeededAndDoOp(GenericCallback<PerChannelBookieClient> op) {
                     // connection attempt either fails or succeeds
                     pendingOps.add(op);
 
-                    if (state == ConnectionState.CONNECTING) {
+                    if (state == ConnectionState.CONNECTING
+                        || state == ConnectionState.STARTTLS) {
                         // just return as connection request has already send
```
Comment (kishorekasi) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/PerChannelBookieClient.java]: reword comment

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java (hunk 12 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/AbstractConfiguration.java]: the option will be reflected to SSLEngine#setEnableProtocols.
I will update the comments on setters and here

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java (hunk 15 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java]: I think this is a part of a previous commit, that has already been merged on master.
The idea is that the AuthProvider will have access to the Certificates, in order to provide custom banning/revoking and automatic close of connection after certificate expiration

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java (hunk 17 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ServerConfiguration.java]: I have no strong opinion on this. Usually other common libraries use properties to store directly passwords and no path to files.
If you need a path to file I can change it.

b.q.
will the access to an absolute file system path  lead to other security concerns ?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java (hunk 18 above)
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/auth/TestAuth.java]: I have added:
- TestSSL#testConnectToSSLClusterMixedClient
- TestSSL#testClientWantsButOnlySSLSomeServerHaveIt

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java (hunk 15 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/auth/BookieAuthProvider.java]: the AuthProvider on this callback has access to the Certificates.
The AuthProvider is instantiated when the connection is created, but no StartTLS as occurred yet, so at that time it is not possible to access client-side Certificates.
So this callback is useful to intercept the client SSL data and:
- schedule a close of the connection at the time of certificate expiration date
- prevent "forbidden" certificates to access Bookie data

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java (hunk 13 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/ssl/SSLContextFactory.java]: IMHO I think that check be done in the AuthProvider.

But we can take care of this directly here.
What do you think ?

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java (hunk 11 above)
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: IMHO I think that check be done in the AuthProvider.

But we can take care of this directly here.
What do you think ?

**Comment by eolivelli** [comment]
@ kishorekasi  thank you for your review.
I have addressed most of the issues and rebased to current master.

I have not run any benchmark on SSL, so I have no number to share about performance degradation

**Comment by sijie** [comment]
@kishorekasi do you mind reviewing the new patch again?

**Comment by sijie** [comment]
@eolivelli since netty 4 is in, do you mind rebasing this pull request?

**Comment by eolivelli** [comment]
@sijie I can rebase, but @kishorekasi told he would have picked this issue up in order to implement more security features.
I am ok for rebasing, getting this merged and then let Kishore or @jvrao create another issue to complete the feature.
A new issue for docs needs to be created as well

**Comment by sijie** [comment]
@kishorekasi @jvrao ^^ any thoughts?

**Comment by eolivelli** [comment]
@sijie @jvrao @kishorekasi 
I will wait for your ACK before starting the rebase I think I will not be trivial

**Comment by kishorekasi** [comment]
Enrico,

I have generate a pull request combing both our changes.

Kishore

**Comment by kishorekasi** [comment]
Sorry, typo. I will generate a pull request.

**Comment by eolivelli** [comment]
@kishorekasi this sounds great to me, I will close this PR once you open yours

**Comment by sijie** [comment]
@kishorekasi any thoughts about the new SSL pull requests?

**Comment by kishorekasi** [comment]
Hi Sijie,

I have some more testing to complete. I will work on it this weekend and post a pull request.

Kishore

**Comment by eolivelli** [comment]
closing this PR in favour of #183 from @kishorekasi

**Pull request closed without merging**